QueueDown/MultiStagePipes.cs

```diff
@@ -0,0 +1,146 @@
+
+
+                // Write to the pipe
+                while (true) {
+                    var buffer = producerPipe.Writer.GetMemory();
+                    buffer.Span.Fill((byte)i);
+                    producerPipe.Writer.Advance(buffer.Length);
+                    await producerPipe.Writer.FlushAsync();
+                }
+            });
+        }
+
+        var _ = Task.Run(() => ReadAllPipes(readers, outputPipe.Writer));
+    }
+
+    /// <summary>
+    /// Reads from pipes in a round-robin manner and pushes output into final pipe
+    /// </summary>
+    private static async Task ReadAllPipes(PipeGroup stage1Pipes, PipeWriter outputPipe)
+    {
+        while (true) {
+            var pipes = stage1Pipes.Readers; //fresh read incase this changed
+            bool anyDataThisLoop = false;
+
+            foreach (var inputPipe in pipes) 
+            {
+                if (inputPipe.TryRead(out var result)) 
+                {
+                    anyDataThisLoop = true;
+
+                    await ProcessPipeResult(stage1Pipes, inputPipe, result, outputPipe);
+                }
+            }
+
+            if(!anyDataThisLoop)
+            {
+                //take the slow path if we are starved
+                var results = await stage1Pipes.WaitForMoreData();
+
+                foreach(var result in results)
+                {
+                    await ProcessPipeResult(stage1Pipes, result.Item1, result.Item2, outputPipe);
+                }
+            }
+        }
+    }
+
+    private static async Task ProcessPipeResult(PipeGroup group, PipeReader inputPipe, ReadResult result, PipeWriter outputPipe)
+    {
+        var inputBuffer = result.Buffer;
+
+        if (inputBuffer.IsSingleSegment) 
+        {
+            outputPipe.Write(result.Buffer.FirstSpan);
+        } 
+        else 
+        {
+            foreach (var segment in result.Buffer) 
+            {
+                outputPipe.Write(segment.Span);
+            }
+        }
+
+        var flushResult = await outputPipe.FlushAsync();
+
+        inputPipe.AdvanceTo(inputBuffer.End);
+
+
+        if (result.IsCompleted || flushResult.IsCompleted) {
+            group.RemovePipe(inputPipe);
+        }
+    }
+
+    
+
+    /// <summary>
+    /// Provides a way to add and remove pipes in a fast/thread safe way
+    /// </summary>
+    public class PipeGroup
+    {
+        private ImmutableArray<PipeReader> _readers = ImmutableArray<PipeReader>.Empty;
+
+        public ImmutableArray<PipeReader> Readers => _readers;
+
+
+        public void AddPipe(PipeReader reader)
+        {
+            ImmutableInterlocked.Update(ref _readers, static (array, arg) => array.Add(arg), reader);
+        }
+
+        public void RemovePipe(PipeReader reader)
+        {
+            ImmutableInterlocked.Update(ref _readers, static (array, arg) => array.Remove(arg), reader);
+        }
+
+        public async Task<IReadOnlyList<(PipeReader, ReadResult)>> WaitForMoreData()
```

Comment:
Allocations! You don't see it now because data is always being written and read. There is no backpressure in this example.

Reply:
Hmm, so we need to avoid allocs along the slow path also. Lemme think on this

Reply:
Unless I'm mistaken, `WaitForMoreData` should only be called during input starvation. This method shouldn't get called if the output/destination pipe is under backpressue. In that case the call to FlushAsync should block. Right?

Reply:
That's right. I misspoke, this technique works great when there's lots of data, but ends up on this code path if streams/requests aren't constantly writing. Say you had few stream(s) writing once per second, you could end up in this code path a lot. That said, this isn't a reason to merge it 😄

Reply:
I guess the other problem is that it doesn't round robin currently so it's really easy to starve the queues. Ideally you would push the change where you start from every iteration to avoid this.